development/gcp/cloud-run/kms-rotate/kms-rotate_test.go

```diff
@@ -0,0 +1,359 @@
+					http.Error(w, "unknown part content type", http.StatusBadRequest)
+					return
+				}
+			}
+			s.files[parsedJSONdata.Name].keyVersion = newData
+
+		}
+	default:
+		{
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, "failed to parse response body: "+path, 500)
+			}
+			fmt.Printf("unknown url: %s ;Query: %s ;Body: %s; %s\n", r.URL.Path, r.URL.Query().Encode(), string(body), r.Method)
+			http.Error(w, "unknown URL: "+path, 500)
+			s.unknownEndpointCallCount++
+		}
+	}
+}
+
+func TestRotateKMSKey(t *testing.T) {
+
+	//var err error
+	ctx := context.Background()
+
+	defaultRequestBody := Config{
+		Project:    testProjectName,
+		Location:   testLocation,
+		BucketName: testBucketName,
+		Keyring:    testKeyringName,
+		Key:        testKeyName,
+	}
+
+	defaultPrefixedRequestBody := Config{
+		Project:      testProjectName,
+		Location:     testLocation,
+		BucketName:   testBucketName,
+		Keyring:      testKeyringName,
+		Key:          testKeyName,
+		BucketPrefix: testPrefix,
+	}
+
+	tests := []struct {
+		name                string
+		requestBody         Config
+		keyVersions         map[string]*fakeKeyVersion
+		expectedKeyVersions map[string]*fakeKeyVersion
+		files               map[string]*fakeFile
+	}{
+		{
+			name:                "One enabled key, no files",
+			requestBody:         defaultRequestBody,
+			keyVersions:         map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_ENABLED}},
+			expectedKeyVersions: map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_ENABLED}},
+		},
+		{
+			name:                "Two enabled keys, no files",
+			requestBody:         defaultRequestBody,
+			keyVersions:         map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_ENABLED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+			expectedKeyVersions: map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_DESTROY_SCHEDULED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+		},
+		{
+			name:                "Two enabled keys, one file",
+			requestBody:         defaultRequestBody,
+			keyVersions:         map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_ENABLED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+			expectedKeyVersions: map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_DESTROY_SCHEDULED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+			files:               map[string]*fakeFile{"cert": {"1"}},
+		},
+		{
+			name:                "Two enabled keys, one file, query with prefix",
+			requestBody:         defaultPrefixedRequestBody,
+			keyVersions:         map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_ENABLED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+			expectedKeyVersions: map[string]*fakeKeyVersion{"1": {kmspb.CryptoKeyVersion_DESTROY_SCHEDULED}, "2": {kmspb.CryptoKeyVersion_ENABLED}},
+			files:               map[string]*fakeFile{"notcert": {"data"}, "certificates/cert": {"1"}},
+		},
+	}
+
+	for _, test := range tests {
```

Comment:
This is also quite long piece of code. Maybe you could try to extract some separate methods from it?